sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -2079,4 +2078,661 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
+      val intPartPos = closeOnExcept(decTemp) { _ =>
+        withResource(intPart) { _ =>
+          withResource(Scalar.fromString("-")) { negativeSign =>
+            withResource(Scalar.fromString("")) { emptyString =>
+              intPart.stringReplace(negativeSign, emptyString)
+            }
+          }
+        }
+      }
+      resource_array += intPartPos
+      val appendZeros = "0" * appendZeroNum
+      val appendZerosCv = closeOnExcept(decTemp) { _ =>
+        withResource(Scalar.fromString(appendZeros)) { zeroString =>
+          ColumnVector.fromScalar(zeroString, cv.getRowCount.toInt)
+        }
+      }
+      val decPart = withResource(decTemp) { _ =>
+        withResource(appendZerosCv) { _ =>
+          ColumnVector.stringConcatenate(Array(decTemp, appendZerosCv))
+        }
+      }
+      (intPartPos, decPart)
+    }
+  }
+
+  private def getParts(cv: ColumnVector, d: Int): (ColumnVector, ColumnVector) = {
+    // get int part and dec part from a column vector, int part will be set to positive
+    x.dataType match {
+      case FloatType | DoubleType => {
+        getPartsFromDouble(cv, d)
+      }
+      case DecimalType.Fixed(_, scale) => {
+        getPartsFromDecimal(cv, d, scale)
+      }
+      case IntegerType | LongType | ShortType | ByteType => {
+        val intPartPos = withResource(cv.castTo(DType.STRING)) { intPart =>
+          withResource(Scalar.fromString("-")) { negativeSign =>
+            withResource(Scalar.fromString("")) { emptyString =>
+              intPart.stringReplace(negativeSign, emptyString)
+            }
+          }
+        }
+        val dzeros = "0" * d
+        val decPart = closeOnExcept(intPartPos) { _ =>
+          withResource(Scalar.fromString(dzeros)) { zeroString =>
+            ColumnVector.fromScalar(zeroString, cv.getRowCount.toInt)
+          }
+        }
+        (intPartPos, decPart)
+      }
+      case _ => {
+        throw new UnsupportedOperationException(s"format_number doesn't support type ${x.dataType}")
+      }
+    }
+  }
+
+  private def negativeCheck(cv: ColumnVector): ColumnVector = {
+    withResource(cv.castTo(DType.STRING)) { cvStr =>
+      withResource(Scalar.fromString("-")) { negativeSign =>
+        cvStr.startsWith(negativeSign)
+      }
+    }
+  }
+
+  private def removeExtraCommas(str: ColumnVector): ColumnVector = {
+    withResource(Scalar.fromString(",")) { comma =>
+      str.rstrip(comma)
+    }
+  }
+
+  private def addCommas(str: ColumnVector): ColumnVector = {
+    val maxstrlen = withResource(str.getCharLengths()) { strlen =>
+      withResource(strlen.max()) { maxlen =>
+        maxlen.isValid match {
+          case true => maxlen.getInt
+          case false => 0
+        }
+      }
+    }
+    val sepCol = withResource(Scalar.fromString(",")) { sep =>
+      ColumnVector.fromScalar(sep, str.getRowCount.toInt)
+    }
+    val substrs = closeOnExcept(sepCol) { _ =>
+      (0 until maxstrlen by 3).map { i =>
```

Comment:
I think we do not need to reverse and reverse back the input string if we slice string from end to start, something like
```
var curEndsCol: Option[ColumnVector] = strlen

withResource(curEndsCol) { _ =>
  (0 until maxstrlen by 3).safeMap { _ =>
    val startCol = curEndsCol - 3,
    val sub = closeOnExcept(startCol) { _ =>
      str.substring(startCol, curEndsCol)
    }
    curEndsCol.close()
    curEndsCol = startCol
    sub
  }
}
```
You need to do `strlen - 3` in columnar way, e.g.
```
withResource(Scalar.fromInt(3)) { scalar3 =>
   strlen.sub(scalar3)
}
```